File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.py
class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        cache = {}
        
        for word in words:
            if word not in cache:
                cache[word] = 0
            cache[word] += 1
            
        result = 0
        isSame = False
        for key, value in cache.items():
            temp = self.getOther(key)
            if temp == key:
                if value > 1:
                    result += (value // 2) * 4
                if not isSame and value % 2:
                    result += 2
                    isSame = True
            elif temp in cache:
                result += min(value, cache[temp]) * 2
        
        return result
    
    def getOther(self, word):
        return f"{word[1]}{word[0]}"

File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.cpp
class Solution {
public:
    int longestPalindrome(vector<string>& words) {
        unordered_map<string, int> cache;
        int result = 0;

        for (string word: words) {
            string reverseWord = word.substr(1, 1) + word.substr(0, 1);
            if (cache.count(reverseWord) > 0) {
                result += 4;
                cache[reverseWord]--;

                if (cache[reverseWord] == 0) {
                    cache.erase(reverseWord);
                }
                continue;
            }

            if (cache.count(word) == 0) {
                cache[word] = 0;
            }
            cache[word]++;
        }

        for (auto[word, _] : cache) {
            if (word[0] == word[1]) {
                result += 2;
                break;
            }
        }

        return result;
    }
};